File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/prototype/request/HttpRequest.java
// HttpRequest.java

package com.bigboxcode.designpattern.prototype.request;

import java.util.HashMap;

enum RequestType {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE,
}

public class HttpRequest extends Request {

    private String url;

    private RequestType type;
    private HashMap<String, String> header = new HashMap<>();
    private HashMap<String, String> body =new HashMap<>();

    public void setUrl(String url) {
        this.url = url;
    }

    public void setType(RequestType type) {
        this.type = type;
    }

    public void addHeader(String key, String value) {
        this.header.put(key, value);
    }

    public void removeHeader(String key) {
        this.header.remove(key);
    }

    public void addBody(String key, String value) {
        this.body.put(key, value);
    }

    public void removeBody(String key) {
        this.body.remove(key);
    }

    public void send() {
        System.out.println("Sending Request...");
        System.out.println("URL: " + url);
        System.out.println("Type: " + type);
        System.out.println("Headers: " + header);
        System.out.println("Body: " + body);


        // Write functional code to send request
    }

}

File: design-patterns/ts/builder/request/http-request.ts
// http-request.ts

export enum RequestType {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE
}

class HttpRequest {
    private url: string;
    private type: RequestType;
    private header = new Map<string, string>();
    private body = new Map<string, string>();

    constructor(
        url: string,
        type: RequestType,
        header: Map<string, string>,
        body: Map<string, string>
        ) {
        this.url = url;
        this.type = type;
        this.header = header;
        this.body = body;
    }

    public send() {
        console.log("Sending Request...");
        console.log("URL: ", this.url);
        console.log("Type: ", this.type);
        console.log("Header: ", this.header);
        console.log("Body: ", this.body);
        
        // Write code to send the request here
    }

    static Builder = class {
        private url: string = "";
        private type: RequestType = RequestType.GET;
        private header = new Map<string, string>();
        private body = new Map<string, string>();

        public setUrl(url: string) {
            this.url = url;
            return this;
        }

        public setType(type: RequestType) {
            this.type = type;
            return this;
        }

        public setHeader(key: string, value: string) {
            this.header.set(key, value);
            return this;
        }

        public setBody(key: string, value: string) {
            this.body.set(key, value);
            return this;
        }

        public build() {
            return new HttpRequest(this.url, this.type, this.header, this.body);
        }
    }
}

export default HttpRequest;